app/data_providers/item2item.py

```diff
@@ -68,23 +73,38 @@ async def _recommend(self, resolved_id: int, query_filter: Filter, count: int):
                     with_payload=True
                 ))).result
         except UnexpectedResponse as ex:
-            if ex.status_code == 404:
-                # point or collection does not exist
-                # it can happen when a new syndicated article was just added or qdrant state was reset by accident
-                # fallback to returning random articles with the same filter
-                # it should fail if the problem is not the article
-                res = (await self._client.scroll_points(
-                    collection_name=self.collection,
-                    scroll_request=ScrollRequest(
-                        limit=count,
-                        filter=query_filter,
-                        with_vector=False,
-                        with_payload=True
-                    ))).result.points
+            if ex.status_code == 404 and 'Not found: No point with id' in ex.content.decode():
+                # article does not exist in qdrant
+                # fallback to returning random popular articles with the same filter
                 logging.warning(f'Related: article is not found, fallback to Qdrant scroll method. '
                                 f'resolved_id: {resolved_id}, filter: {query_filter}')
+                res = await self._fallback(count, query_filter)
             else:
-                raise
+                logging.error(f'Related: unexpected response from qdrant. '
+                              f'Code: {ex.status_code}, reason: {ex.reason_phrase}')
+                res = []
 
         return [CorpusItemModel(id=rec.payload['corpus_item_id'], topic=rec.payload['topic'])
                 for rec in res]
+
+    async def _fallback(self, count, query_filter):
+        # use frequently saved articles for all queries except publisher based
+        # because there might be not enough saves for some publishers
+        if 'domain' not in {cond.key for cond in query_filter.must}:
+            query_filter.must.append(FieldCondition(key='save_count', range=Range(gte=1000)))
+        res = []
+
+        try:
+            res = (await self._client.scroll_points(
+                collection_name=self.collection,
+                scroll_request=ScrollRequest(
+                    limit=count,
+                    filter=query_filter,
+                    with_vector=False,
+                    with_payload=True
+                ))).result.points
```

Comment:
This fallback is reasonable for Syndicated and Publisher-based queries, but for the `related` query it could return old curated items that aren't relevant anymore. Would it be better not to return anything in that case?

Usually we QA recommendation results, but I don't see a ticket in [TDP-87](https://getpocket.atlassian.net/browse/TDP-87) to do that. Would you like to create a ticket for that?

Reply:
Good question, the assumption is that it's better to show something than nothing, especially if qdrant is down or the article is absent in the vector store which can happen frequently. I will create a ticket for QA.